crypto/stateproof/builder.go

```diff
@@ -0,0 +1,256 @@
+	ErrPositionAlreadyPresent = errors.New("requested position is already present")
+	ErrPositionWithZeroWeight = errors.New("position has zero weight")
+	ErrCoinIndexError         = errors.New("could not find corresponding index for a given coin")
+)
+
+// Builder keeps track of signatures on a message and eventually produces
+// a state proof for that message.
+type Builder struct {
+	data           MessageHash
+	round          uint64
+	sigs           []sigslot // Indexed by pos in participants
+	signedWeight   uint64    // Total weight of signatures so far
+	participants   []basics.Participant
+	parttree       *merklearray.Tree
+	lnProvenWeight uint64
+	provenWeight   uint64
+	strengthTarget uint64
+}
+
+// MkBuilder constructs an empty builder. After adding enough signatures and signed weight, this builder is used to create a stateproof.
+func MkBuilder(data MessageHash, round uint64, provenWeight uint64, part []basics.Participant, parttree *merklearray.Tree, strengthTarget uint64) (*Builder, error) {
+	npart := len(part)
+	lnProvenWt, err := LnIntApproximation(provenWeight)
+	if err != nil {
+		return nil, err
+	}
+
+	b := &Builder{
+		data:           data,
+		round:          round,
+		sigs:           make([]sigslot, npart),
+		signedWeight:   0,
+		participants:   part,
+		parttree:       parttree,
+		lnProvenWeight: lnProvenWt,
+		provenWeight:   provenWeight,
+		strengthTarget: strengthTarget,
+	}
+
+	return b, nil
+}
+
+// Present checks if the builder already contains a signature at a particular
+// offset.
+func (b *Builder) Present(pos uint64) (bool, error) {
+	if pos >= uint64(len(b.sigs)) {
+		return false, fmt.Errorf("%w pos %d >= len(b.sigs) %d", ErrPositionOutOfBound, pos, len(b.sigs))
+	}
+
+	return b.sigs[pos].Weight != 0, nil
+}
+
+// IsValid verifies that the participant along with the signature can be inserted to the builder.
+// verifySig can be set to false when the signature is already verified (e.g. loaded from the DB)
+func (b *Builder) IsValid(pos uint64, sig *merklesignature.Signature, verifySig bool) error {
+	if pos >= uint64(len(b.participants)) {
+		return fmt.Errorf("%w pos %d >= len(participants) %d", ErrPositionOutOfBound, pos, len(b.participants))
+	}
+
+	p := b.participants[pos]
+
+	if p.Weight == 0 {
+		return fmt.Errorf("builder.IsValid: %w: position = %d", ErrPositionWithZeroWeight, pos)
+	}
+
+	// Check signature
+	if verifySig {
+		if err := sig.ValidateSaltVersion(merklesignature.SchemeSaltVersion); err != nil {
+			return err
+		}
+
+		cpy := make([]byte, len(b.data))
+		copy(cpy, b.data[:]) // TODO: once cfalcon is fixed can remove this copy.
+		if err := p.PK.VerifyBytes(b.round, cpy, sig); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// Add a signature to the set of signatures available for building a proof.
+func (b *Builder) Add(pos uint64, sig merklesignature.Signature) error {
+	isPresent, err := b.Present(pos)
```

Comment:
The `Add` function must protect itself since If we try to add the same signature twice it would be wrong since the signed weight would be increased twice. 
Regrading `IsValid`  function (the function above). it mostly checks that the signature is valid. 
it should be a separate function since the reaction on an invalid signature and a signature that already exists should be different.